File: MIPS/test/assembler.cpp
#include <bits/stdc++.h>

#include <iostream>
#include <sstream>
#include <fstream>
#include <vector>
#include <string>
#include <unordered_map>
#include <regex>
#include <bitset>
#include <iomanip>
using namespace std;

class MIPSAssembler
{
private:
    //   R-type instruction opcodes and function codes
    std::unordered_map<std::string, std::string> r_type_opcodes = {
        {"nor", "0101"}, {"and", "1010"}, {"or", "1100"}, {"srl", "0111"}, {"sll", "1000"}, {"add", "0100"}, {"sub", "1001"}};

    // I-type instruction opcodes
    std::unordered_map<std::string, std::string> i_type_opcodes = {
        {"subi", "0010"}, {"sw", "0001"}, {"bneq", "1110"}, {"andi", "0110"}, {"addi", "1011"}, {"ori", "0011"}, {"lw", "0000"}, {"beq", "1101"}};

    // J-type instruction opcodes
    std::unordered_map<std::string, std::string> j_type_opcodes = {
        {"j", "1111"}};

    // Register mapping
    std::unordered_map<std::string, std::string> registers = {
        {"$zero", "0000"},
        {"$t0", "0001"},
        {"$t1", "0010"},
        {"$t2", "0011"},
        {"$t3", "0100"},
        {"$t4", "0101"},
        {"$sp", "0110"},
        {"$ra", "0111"}};

    // Helper function to convert decimal to binary with specified bit length
    std::string decimalToBinary(int decimalNum, int bits)
    {
        // Handle negative numbers using two's complement
        if (decimalNum < 0)
        {
            decimalNum = (1 << bits) + decimalNum;
        }

        // Convert to binary string with leading zeros
        std::bitset<16> binary(decimalNum);
        std::string binaryStr = binary.to_string();
        return binaryStr.substr(binaryStr.length() - bits);
    }

public:
    // Parse a single instruction and convert to machine code
    std::string parseInstruction(const std::string &instruction)
    {
        // Remove comments and trim whitespace
        std::string cleanInstr = instruction;
        size_t commentPos = cleanInstr.find("#");
        if (commentPos != std::string::npos)
        {
            cleanInstr = cleanInstr.substr(0, commentPos);
        }

        // Trim leading and trailing whitespace
        cleanInstr.erase(0, cleanInstr.find_first_not_of(" \t"));
        cleanInstr.erase(cleanInstr.find_last_not_of(" \t") + 1);

        if (cleanInstr.empty())
        {
            return "";
        }

        // Split instruction into parts
        std::vector<std::string> parts;
        std::istringstream iss(cleanInstr);
        std::string part;
        while (std::getline(iss, part, ','))
        {
            std::istringstream partStream(part);
            while (partStream >> part)
            {
                parts.push_back(part);
            }
        }

        std::string opcode = parts[0];
        std::transform(opcode.begin(), opcode.end(), opcode.begin(), ::tolower);

        // R-type instructions
        if (r_type_opcodes.count(opcode) > 0)
        {
            std::string inst_opcode = r_type_opcodes[opcode];
            std::string rd = registers[parts[1]];
            std::string rs = registers[parts[2]];
            std::string rt = registers[parts[3]];

            string instruction = inst_opcode + rs + rt + rd + "0000";

            if (inst_opcode == "0111" || inst_opcode == "1000")
                instruction = inst_opcode + rs + "1000" + rd + decimalToBinary(stoi(parts[3]), 4);

            return instruction;
        }
        // I-type instructions
        else if (i_type_opcodes.count(opcode) > 0)
        {
            std::string inst_opcode = i_type_opcodes[opcode];
            std::string rs, rt, immediate;

            // Memory instructions (lw, sw)
            if (opcode == "lw" || opcode == "sw")
            {
                std::regex memoryPattern(R"(\$(\w+),\s*(-?\d+)\((\$\w+)\))");
                std::smatch match;
                std::string fullMemoryPart = parts[1] + ", " + parts[2];

                if (std::regex_search(fullMemoryPart, match, memoryPattern))
                {
                    rt = registers[std::string("$") + match.str(1)];
                    rs = registers[match.str(3)];
                    immediate = decimalToBinary(std::stoi(match.str(2)), 8);
                }
                else
                {
                    return "Invalid memory instruction";
                }
            }
            // Other I-type instructions
            else
            {
                rs = registers[parts[2]];
                rt = registers[parts[1]];
                immediate = decimalToBinary(std::stoi(parts[3]), 8);
            }

            return inst_opcode + rs + rt + immediate;
        }

        return "Invalid instruction: " + cleanInstr;
    }

    // Convert entire assembly program to machine code
    std::vector<std::string> convertProgram(const std::string &assemblyCode)
    {
        std::vector<std::string> machineCode;
        std::istringstream stream(assemblyCode);
        std::string line;

        while (std::getline(stream, line))
        {
            std::string machineInstruction = parseInstruction(line);
            if (!machineInstruction.empty() && machineInstruction.find("Invalid") == std::string::npos)
            {
                // Convert binary to hex
                std::bitset<20> bits(machineInstruction);
                std::stringstream ss;
                ss << std::hex << std::setfill('0') << std::setw(5) << bits.to_ulong();
                machineCode.push_back(ss.str());
            }
        }

        return machineCode;
    }
};

int main()
{
    MIPSAssembler assembler;
    std::string assemblyCode;
    std::ifstream inputFile("Input.txt");

    if (!inputFile)
    {
        std::cerr << "Error opening input file" << std::endl;
        return 1;
    }

    // Read entire file into string
    assemblyCode.assign((std::istreambuf_iterator<char>(inputFile)),
                        std::istreambuf_iterator<char>());

    // Convert assembly to machine code
    std::vector<std::string> machineCode = assembler.convertProgram(assemblyCode);

    // Write machine code to output file
    std::ofstream outputFile("Output.txt");
    if (!outputFile)
    {
        std::cerr << "Error opening output file" << std::endl;
        return 1;
    }

    // Write header for LogiSim compatibility
    outputFile << "v2.0 raw\n";
    for (const auto &code : machineCode)
    {
        outputFile << code << "\n";
    }

    return 0;
}

File: MIPS/test/assembler.py
"""
Assembler code to convert mips assembly code to machine code
"""

import re
import sys

instruction_opcode = {
    'lw': '0', 'sw': '1', 'subi': '2', 'ori': '3', 'add': '4', 'nor': '5', 'andi': '6', 'srl': '7',
    'sll': '8', 'sub': '9', 'and': 'a', 'addi': 'b', 'or': 'c', 'beq': 'd', 'bneq': 'e', 'j': 'f'
}


registers = {'$zero': '0', '$t0': '1', '$t1': '2', '$t2': '3',
             '$t3': '4', '$t4': '5', '$sp': '6', '$ra': '7'}


def isRInstruction(instruction):
    return instruction in ['nor', 'and', 'or', 'add', 'sub']


def isMemoryInstruction(instruction):
    return instruction in ['sw', 'lw']


def isBranchInstruction(instruction):
    return instruction in ['beq', 'bneq']


def isShiftInstruction(instruction):
    return instruction in ['srl', 'sll']


def isImmidiateInstruction(instruction):
    return instruction in ['subi', 'andi', 'addi', 'ori']


def isJumpInstruction(instruction):
    return instruction in ['j']


label_found = 0
label_dict = {}

instruction_list = []
machine_code_list = []

input_file = sys.argv[1]
output_file = sys.argv[2]

with open(input_file, 'r', encoding='ascii') as f:
    for line in f:
        without_comment = re.split(r'[#]+', line)[0].strip()
        if without_comment:
            instruction_list.append(without_comment.strip())
            print(instruction_list[-1])

for i, instruction in enumerate(instruction_list):
    if ':' in instruction:
        label = instruction.split(':')[0]
        label_dict[label] = i - label_found
        label_found += 1

for i, instruction in enumerate(instruction_list):
    if ':' in instruction:
        instruction_list.pop(i)

for i, instruction in enumerate(instruction_list):
    opcode = instruction.split(' ')[0]
    if isRInstruction(opcode):
        parts = re.split(r'[ ,]+', instruction)
        machine_code = instruction_opcode[opcode] + \
            registers[parts[2]] + registers[parts[3]] + \
            registers[parts[1]] + '0'
        machine_code_list.append(machine_code)

    elif isImmidiateInstruction(opcode):
        parts = re.split(r'[ ,]+', instruction)
        immidiate_value = int(parts[3])
        immidiate_value = (
            1 << 8) + immidiate_value if immidiate_value < 0 else immidiate_value

        machine_code = instruction_opcode[opcode] + \
            registers[parts[2]] + registers[parts[1]] + \
            format(immidiate_value, '02x')
        machine_code_list.append(machine_code)

    elif isShiftInstruction(opcode):
        parts = re.split(r'[ ,]+', instruction)
        shift_value = int(parts[3])
        shift_value = (
            1 << 4) + shift_value if shift_value < 0 else shift_value
        machine_code = instruction_opcode[opcode] + \
            registers[parts[2]] + '0'+registers[parts[1]] + \
            format(shift_value, '01x')
        machine_code_list.append(machine_code)

    elif isMemoryInstruction(opcode):
        parts = re.split(r'[ ,()]+', instruction)
        offset = int(parts[2])
        offset = (
            1 << 8) + offset if offset < 0 else offset
        machine_code = instruction_opcode[opcode] + \
            registers[parts[3]] + registers[parts[1]] + \
            format(offset, '02x')
        machine_code_list.append(machine_code)

    elif isBranchInstruction(opcode):
        parts = re.split(r'[ ,]+', instruction)
        position = label_dict[parts[3]] - i - 1
        position = (1 << 8) + position if position < 0 else position
        machine_code = instruction_opcode[opcode] + \
            registers[parts[1]] + registers[parts[2]] + \
            format(position, '02x')
        machine_code_list.append(machine_code)

    elif isJumpInstruction(opcode):
        parts = re.split(r'[ ,]+', instruction)
        position = label_dict[parts[1]]
        machine_code = instruction_opcode[opcode] + \
            format(position, '02x') + '00'
        machine_code_list.append(machine_code)


machine_code_list.append('e0000')
machine_code_list.append('e0000')
machine_code_list.append('e0000')
machine_code_list.append('e0000')

with open(output_file, 'w', encoding='ascii') as f:
    f.write('v2.0 raw\n')
    for machine_code in machine_code_list:
        f.write(machine_code + '\n')
